src/checkers/inference/model/serialization/ToStringSerializer.java

```diff
@@ -24,49 +28,79 @@
 import checkers.inference.model.VariableSlot;
 
 /**
- * This Serializer is meant only to convert constraints and variables to
- * human readable strings.  It is used currently by the DebugSolver.
+ * This Serializer converts constraints and variables to human readable strings.
  */
 public class ToStringSerializer implements Serializer<String, String> {
     private final boolean showAstPaths;
     private boolean showVerboseVars;
 
-    private int indent = 0;
-    public static final String INDENT_STRING = "    ";
+    /**
+     * Stores the current indentation level, where level 0 means no indentations at all, level 1
+     * means 1 indentation and so on. Each indentation is equivalent to the number of spaces stored
+     * in {@link #INDENT}.
+     */
+    private int indentationLevel = 0;
+
+    /**
+     * This string constant stores the number of spaces per indentation. It is set to 2 spaces per
+     * indentation.
+     */
+    private static final String INDENT = "  ";
+
+    /**
+     * This list holds the indent strings for each indentation level, where it stores N
+     * concatenations of INDENT at index N. The index is the indentation level. Index 0 stores empty
+     * string.
+     */
+    private final List<String> indentStrings = new ArrayList<>();
+
+    // used to format constant slots
+    private final AnnotationFormatter formatter;
 
     public ToStringSerializer(boolean showAstPaths) {
         this.showAstPaths = showAstPaths;
         this.showVerboseVars = true;
-    }
+        formatter = InferenceMain.getInstance().getRealTypeFactory().getAnnotationFormatter();
 
-    public void setIndent(int indent) {
-        this.indent = indent;
+        // set first value to ""
+        indentStrings.add("");
     }
 
-    public int getIndent() {
-        return indent;
+    public void setIndentationLevel(int indentationLevel) {
+        // Ensure the indentation level is always >= 0
+        this.indentationLevel = indentationLevel > 0 ? indentationLevel : 0;
+        // create additional indentation strings up to the current level of indentation, if it
+        // doesn't exist in the indentStrings array list
+        // subsequent indentation string values are 1 more INDENT compared to the previous index
+        StringBuilder sb = new StringBuilder(indentStrings.get(indentStrings.size() - 1));
```

Comment:
If we set indentation level to old indentation level - 1, `sb` is still created, but isn't used.

Reply:
added an if block around the updates to indentStrings list